From f2849b1fd059ec9b3281b771e6ac5aad9feee851 Mon Sep 17 00:00:00 2001
From: Christian Marangi <ansuelsmth@gmail.com>
Date: Mon, 17 Aug 2026 23:30:03 +0200
Subject: [PATCH] net: phylink: correctly validate returned PCS in
 phylink_inband_caps

commit f2849b1fd059ec9b3281b771e6ac5aad9feee851 upstream.

[ Vanilla 6.18.54 already carries this fix (stable backport), but the
  pending 737-02 "net: phylink: introduce internal phylink PCS handling"
  rewrites phylink_inband_caps() and reintroduces a plain NULL check
  (pcs_found = !!pcs). Re-apply the IS_ERR_OR_NULL() check on top of it.
  Alternatively fold this hunk into 737-02. ]

In phylink_inband_caps(), the PCS returned by mac_select_pcs is only
checked if NULL but mac_select_pcs can also return an error pointer.

This can cause a kernel panic as phylink_pcs_inband_caps() only checks
if passed PCS is not NULL and directly dereference ops from the phylink_pcs
struct.

Use the IS_ERR_OR_NULL macro to address both case where the returned
PCS can be NULL or an error pointer and prevent a kernel panic.

Cc: stable@vger.kernel.org
Fixes: df874f9e52c3 ("net: phylink: add pcs_inband_caps() method")
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>
Link: https://patch.msgid.link/20260817213009.13924-1-ansuelsmth@gmail.com
Signed-off-by: Jakub Kicinski <kuba@kernel.org>
---
 drivers/net/phy/phylink.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/net/phy/phylink.c b/drivers/net/phy/phylink.c
index 57c402beb..54b8fc4bd 100644
--- a/drivers/net/phy/phylink.c
+++ b/drivers/net/phy/phylink.c
@@ -1020,7 +1020,7 @@ static unsigned int phylink_inband_caps(struct phylink *pl,
 	if (pl->mac_ops->mac_select_pcs) {
 		pcs = pl->mac_ops->mac_select_pcs(pl->config,
 						  interface);
-		pcs_found = !!pcs;
+		pcs_found = !IS_ERR_OR_NULL(pcs);
 	} else if (test_bit(interface, pl->config->pcs_interfaces)) {
 		list_for_each_entry(pcs, &pl->pcs_list, list) {
 			if (!phylink_validate_pcs_interface(pcs, interface)) {
